### PR TITLE
refactor: refactor BufferedWriter

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

- Refactor the `BufferedWriter` and move it to `datasource` crate.
- Add `JsonEncoder` and `CsvEncoder`.

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
- Sub-task of #1305

### CODE REVIEW

**Comment by v0y4g3r** [comment]
We've already had a similar struct [`BufferedWriter`](https://github.com/GreptimeTeam/greptimedb/blob/0253136333dd8deb2fda05b66dde077da6e5d8a0/src/storage/src/sst/stream_writer.rs#L63) which buffers encoded record batch in memory until a threshold and then append to opendal.

**Comment by WenyXu** [comment]
> We've already had a similar struct [`BufferedWriter`](https://github.com/GreptimeTeam/greptimedb/blob/0253136333dd8deb2fda05b66dde077da6e5d8a0/src/storage/src/sst/stream_writer.rs#L63) which buffers encoded record batch in memory until a threshold and then append to opendal.

I saw it. Should I refactor the `BufferedWriter` based on my implementation? Currently, the `BufferedWriter` only supports  `ArrowWriter`.

**Comment by WenyXu** [comment]
PTAL @v0y4g3r @waynexia @evenyag

File: src/common/datasource/src/buffered_writer.rs [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    pub async fn close_with_arrow_writer(mut self) -> Result<(FileMetaData, u64)> {
+        if let Some(encoder) = self.encoder.take() {
+            let metadata = encoder.close().await?;
+            let written = self.try_flush(true).await?;
+
+            // It's important to shut down! flushes all pending writes
+            self.close().await?;
+            return Ok((metadata, written));
+        }
+        error::BufferedWriterClosedSnafu {}.fail()
```
Comment (v0y4g3r) [inline src/common/datasource/src/buffered_writer.rs:59]:
```suggestion
        let encoder = self
            .encoder
            .take()
            .context(error::BufferedWriterClosedSnafu)?;
        let metadata = encoder.close().await?;
        let written = self.try_flush(true).await?;

        // It's important to shut down! flushes all pending writes
        self.close().await?;
        return Ok((metadata, written));
```

File: src/storage/src/sst/stream_writer.rs [hunk 2]
```diff
@@ -12,65 +12,31 @@
...
 /// storage by chunks to reduce memory consumption.
 pub struct BufferedWriter {
-    path: String,
-    arrow_writer: ArrowWriter<Buffer>,
-    object_writer: Writer,
-    buffer: Buffer,
-    bytes_written: AtomicU64,
-    flushed: bool,
-    threshold: usize,
+    inner: Box<InnerBufferedWriter>,
```
Comment (v0y4g3r) [inline src/storage/src/sst/stream_writer.rs:34]: Looks like there's no need to box it.

File: src/common/datasource/src/buffered_writer.rs [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        let remain = self.buffer.buffer.lock().unwrap().split();
+        let size = remain.len();
+
+        self.writer
+            .write_all(&remain)
+            .await
+            .context(error::AsyncWriteSnafu)?;
+        self.flushed = true;
+
+        self.bytes_written += size as u64;
```
Comment (MichaelScofield) [inline src/common/datasource/src/buffered_writer.rs:138]: duplicated with line 123?

Comment (WenyXu) [reply src/common/datasource/src/buffered_writer.rs:138]: However, It flushes the non-aligned(the remaining) chunk to the underlying writer.

File: src/common/datasource/src/buffered_writer.rs [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+                buffer.split_to(self.threshold)
+            };
+            let size = chunk.len();
+
+            self.writer
+                .write_all(&chunk)
+                .await
+                .context(error::AsyncWriteSnafu)?;
+
+            self.flushed = true;
```
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/common/datasource/src/buffered_writer.rs:115]: I don't understand the semantics of `flushed` in the context, I have no idea whether we should change this, which may break the original design potentially. @v0y4g3r

File: src/common/datasource/src/buffered_writer.rs [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    pub fn flushed(&self) -> bool {
+        self.flushed
+    }
+
+    pub async fn try_flush(&mut self, all: bool) -> Result<u64> {
+        let mut bytes_written: u64 = 0;
+
+        // Once buffered data size reaches threshold, split the data in chunks (typically 4MB)
+        // and write to underlying storage.
+        while self.buffer.buffer.lock().unwrap().len() >= self.threshold {
```
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/common/datasource/src/buffered_writer.rs:105]: We need to ensure the guard was dropped before the calling await.

File: src/common/datasource/src/buffered_writer.rs [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+    pub async fn write(&mut self, batch: &RecordBatch) -> Result<()> {
+        if let Some(encoder) = &mut self.encoder {
+            encoder.write(batch)?;
+
+            self.try_flush(false).await?;
+
+            return Ok(());
+        }
+        error::BufferedWriterClosedSnafu {}.fail()
```
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/common/datasource/src/buffered_writer.rs]: I tried to remove this by using context. However, I got the following:
```
`snafu::OptionExt::context` takes ownership of the receiver `self`, which moves `self.encoder`
```

[the remark this answers is not in the document]
Comment (v0y4g3r) [reply src/common/datasource/src/buffered_writer.rs]: > I tried to remove this by using context. However, I got the following:
> 
> ```
> `snafu::OptionExt::context` takes ownership of the receiver `self`, which moves `self.encoder`
> ```

```rs
    pub async fn write(&mut self, batch: &RecordBatch) -> Result<()> {
        let encoder = self
            .encoder
            .as_mut()
            .context(error::BufferedWriterClosedSnafu)?;
        encoder.write(batch)?;
        self.try_flush(false).await?;
        return Ok(());
    }

```

File: src/common/datasource/src/buffered_writer.rs (hunk 3 above)
Comment (WenyXu) [reply src/common/datasource/src/buffered_writer.rs:136]: > However, It flushes the non-aligned(the remaining) chunk to the underlying writer.

I Moved it to `try_flush`

File: src/common/datasource/src/buffered_writer.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/common/datasource/src/buffered_writer.rs]: Thx @v0y4g3r .

I misunderstood the original comment. I double-checked this part and moved this line to the bottom of the function.

```rust
...
            let size = chunk.len();

            self.writer
                .write_all(&chunk)
                .await
                .context(error::AsyncWriteSnafu)?;

            bytes_written += size as u64;
        }

        if all {
            bytes_written += self.try_flush_all().await?;
        }

        self.flushed = bytes_written > 0;
        self.bytes_written += bytes_written;

        Ok(bytes_written)
    }
```

**Comment by WenyXu** [comment]
PTAL @v0y4g3r @MichaelScofield

**Review by MichaelScofield — APPROVED** [review]

File: src/common/datasource/src/buffered_writer.rs (hunk 3 above)
Comment (v0y4g3r) [reply src/common/datasource/src/buffered_writer.rs:136]: Does the `bytes_written` accumulated twice?

Comment (WenyXu) [reply src/common/datasource/src/buffered_writer.rs]: > Does the `bytes_written` accumulated twice?

Oh, you are right. And I fixed it.

**Review by v0y4g3r — APPROVED** [review]
LGTM

**Pull request merged**